stan/math/prim/prob/von_mises_cdf.hpp

```diff
@@ -0,0 +1,124 @@
+                                                  const T_p& p) {
+  using stan::math::cos;
+  using stan::math::sin;
+
+  double pi = stan::math::pi();
+  auto s = sin(x);
+  auto c = cos(x);
+  auto sn = sin(p * x);
+  auto cn = cos(p * x);
+
+  return_type_t<T_x, T_k, T_p> R = 0;
+  return_type_t<T_x, T_k, T_p> V = 0;
+
+  int n;
+  for (n = p - 1; n > 0; n--) {
+    auto sn_tmp = sn * c - cn * s;
+    cn = cn * c + sn * s;
+    sn = sn_tmp;
+    R = 1 / (2.0 * n / k + R);
+    V = R * (sn / n + V);
+  }
+  return 0.5 + x / (2 * pi) + V / pi;
+}
+
+template <typename T_x, typename T_k>
+return_type_t<T_x, T_k> von_mises_cdf_normalapprox(const T_x& x, const T_k& k) {
+  using std::exp;
+  using std::sqrt;
+  double pi = stan::math::pi();
+
+  auto b = sqrt(2 / pi) * exp(k) / modified_bessel_first_kind(0, k);
+  auto z = b * sin(x / 2);
+  double mu = 0;
+  double sigma = 1;
+
+  return normal_cdf(z, mu, sigma);
+}
+
+template <typename T_x, typename T_k>
+return_type_t<T_x, T_k> von_mises_cdf_centered(const T_x& x, const T_k& k) {
+  // if the scale is sufficiently small, approximate the cdf with a
+  // normal distribution, otherwise, use the expansion from scipy
+  double ck = 50;
+  double a[4] = {28, 0.5, 100, 5};
+  return_type_t<T_x, T_k> f;
+  if (k < ck) {
+    int p = value_of_rec(a[0] + a[1] * k - a[2] / (k + a[3]) + 1);
+    f = von_mises_cdf_series(x, k, p);
+    if (f < 0)
+      f = 0;
+    if (f > 1)
+      f = 1;
+  } else {
+    f = von_mises_cdf_normalapprox(x, k);
+  }
+  return f;
+}
+
+}  // namespace internal
+
+/** \ingroup prob_dists
+ * Calculates the cumulative distribution function of the von Mises
+ * distribution:
+ *
+ * \f$VonMisesCDF(x, mu, \kappa) = \frac{1}{2\pi I_0(\kappa)} \int_{-\mu-\pi}^x
+ * e^{\kappa cos(t - \mu)} dt\f$
+ *
+ * where
+ *
+ * \f$\mu \in \mathbb{R}\f$, \f$x \in (\mu - \pi, \mu + \pi)\f$, and \f$\kappa
+ * \in \mathbb{R}^+\f$.
+ *
+ * @param x A scalar variate
+ * @param mu The mean of the distribution
+ * @param k The inverse scale of the distriubtion
+ * @return The von Mises cdf evaluated at the specified arguments
+ * @tparam T_x Type of x
+ * @tparam T_mu Type of mean parameter
+ * @tparam T_k Type of inverse scale parameter
+ */
+template <typename T_x, typename T_mu, typename T_k>
+inline return_type_t<T_x, T_mu, T_k> von_mises_cdf(const T_x& x, const T_mu& mu,
+                                                   const T_k& k) {
+  static char const* const function = "von_mises_cdf";
+  using T_partials_return = partials_return_t<T_x, T_mu, T_k>;
+  using internal::von_mises_cdf_centered;
+
+  check_not_nan(function, "Random variable", x);
+  check_not_nan(function, "Scale parameter", k);
+  check_not_nan(function, "Location parameter", mu);
+  check_positive(function, "Scale parameter", k);
+
+  // shift x so that mean is 0
+  return_type_t<T_x, T_mu, T_k> x2 = x - mu;
+
+  // x is on an interval (2*n*pi, (2*n + 1)*pi), move it to (-pi, pi)
+  double pi = stan::math::pi();
```

Comment:
I'm not sure we want the cdf to be periodic. I'm not sure why this doesn't have the support cleanly defined. @bgoodri do you have an opinion here? This seems a bit sketchy. But also saying the support is ```(mu - pi, to mu + pi)``` is weird cause then the support is a function of the parameters.

There's a little note in the docs, but it's also weirdly vague: https://mc-stan.org/docs/2_22/functions-reference/von-mises-distribution.html (starting with: "In order for this density to properly normalize")

Reply:
The docs are a little unclear and also don't reflect what the pdf is actually doing. I made the cdf periodic in order to be consistent with the pdf. We could just define all the von mises functions so that they have mu=0 and support on (-pi, pi).